File: Typescript/Inheritance/Hybrid.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var Person = /** @class */ (function () {
    function Person(name) {
        this.name = name;
    }
    return Person;
}());
var Surgeon = /** @class */ (function (_super) {
    __extends(Surgeon, _super);
    function Surgeon() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Surgeon.prototype.treatPatient = function () {
        console.log("Treating patient");
    };
    Surgeon.prototype.doResearch = function () {
        console.log("Doing medical research");
    };
    Surgeon.prototype.performSurgery = function () {
        console.log("Performing surgery");
    };
    return Surgeon;
}(Person));
var surgeon = new Surgeon("Dr. Kumar");
surgeon.treatPatient();
surgeon.doResearch();
surgeon.performSurgery();

File: c++/Inheritance/Hybrid.cpp
#include <iostream>
using namespace std;

// Base class
class Person {
public:
    void showIdentity() {
        cout << "I am a person" << endl;
    }
};
// Single inheritance
class MedicalStaff : public Person {
public:
    void work() {
        cout << "Medical staff working" << endl;
    }
};
// Multiple inheritance (Level 1)
class Doctor : virtual public MedicalStaff {
public:
    void treatPatient() {
        cout << "Doctor treating patient" << endl;
    }
};
class Nurse : virtual public MedicalStaff {
public:
    void assistDoctor() {
        cout << "Nurse assisting doctor" << endl;
    }
};
// Hybrid inheritance (Multiple + Multilevel)
class Surgeon : public Doctor, public Nurse {
public:
    void performSurgery() {
        cout << "Surgeon performing surgery" << endl;
    }
};

int main() {
    Surgeon s;

    s.showIdentity();      // From Person
    s.work();              // From MedicalStaff
    s.treatPatient();      // From Doctor
    s.assistDoctor();      // From Nurse
    s.performSurgery();    // Own method

    return 0;
}
